java/src/main/java/ai/rapids/cudf/CSVOptions.java

```diff
@@ -91,6 +97,18 @@ public static Builder builder() {
     private int headerRow = NO_HEADER_ROW;
     private byte delim = ',';
     private byte quote = '"';
+    private int rowGuess = -1;
+
+    /**
+     * A guess of how many rows will be loaded. This is totally optional and only used for
+     * estimating the memory usage of loading the data.
+     * @param guess the number of rows to be loaded as a guess.
+     * @return this for chaining.
+     */
+    public Builder withRowGuess(int guess) {
```

Comment:
Does a row count make the most sense to use as a guess here?  Thinking of the string case and other variable-width types that will foil this approach.  Wondering if it should just be a memory size guess to let the caller have more control over the expected size if they happen to know more about the expected row contents than the schema would imply.  cudf could provide a utility method to return the minimum size of a schema (i.e.: just using the fixed-width types in the schema) and the user could control the fudging from there.

Reply:
This is one thing I was thinking about and I am not totally sure what the right thing to do is. We can ask users to provide a size guess which is very generic, but it is partly pushing the burden back onto them.  Do they know the schema before it is loaded?  Possibly. If they do are they aware of approximately how much memory each row will take based off of that schema? You mention strings as an example but I don't think they know any better than us how large the string data really is. The end user may have some knowledge but we have other knowledge and to get a final imperfect guess we have to combine them together.  rows felt like something that the user might be able to come close to guessing at.

Reply:
What I think would work best is having the user specify the memory size estimate.  In the case of the Spark plugin loading Parquet and ORC, the plugin knows far more about what's about to be loaded than the cudf layer does.  Therefore putting the guessing logic in cudf doesn't make sense to me.  The plugin knows the full schema and row count, so it can come up with a fairly close estimate if the schema is just fixed-width types.  This also allows it to control how variable-width types would be estimated (e.g.: controllable via a config settable by the end-user).

If cudf hardcodes the guessing logic then trying to control the memory size estimate on the caller's end becomes messy.  The caller would need to understand how cudf translates row counts to a memory size so the caller can fudge the reported row count accordingly to get the desired size.  If we keep the estimate as a row size then arguably there should be a way to specify the variable-width fields estimate so the caller has better control over the memory estimate when it knows more than cudf does.  At that point I think it's simpler to have only one API to have the caller specify the memory size directly.

> it is partly pushing the burden back onto them.

Yes, but cudf could provide a utility function to calculate a minimum row size for a schema. That would offload some of the calculation burden from the caller and keep the specifics of type sizes within cudf.  The caller would then be responsible for determining the sizes of variable-width fields to get the final average row size, and maybe it has access to some data that helps there.